### PR TITLE
*: fix a bug that wrong index data on prefixed clustered index 

### LINKED ISSUE #23707: insert on char(n) prefix primary key with new collation lead to inconsistent
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

1. setup up 2 tidb instances with new collation on
2. execute following sql on upstream
   ```sql
   create table t (k char(20), v int, primary key (k(4)) /*T![clustered_index] clustered */, key (k)) collate utf8mb4_general_ci;
   insert into t values ('01234', 1), ('12345', 2), ('23456', 1);
   ```
3. backup the table and restore it to downstream
4. do admin check table on downstream

### 2. What did you expect to see? (Required)

No error reported.

### 3. What did you see instead (Required)

```
> admin check table t;
(1105, 'col k, handle &kv.CommonHandle{encoded:[]uint8{0x1, 0x0, 0x30, 0x0, 0x31, 0x0, 0x32, 0x0, 0x33, 0xff, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xf7}, colEndOffsets:[]uint16{0x13}}, index:types.Datum{k:0x5, decimal:0x0, length:0x0, i:0, collation:"utf8mb4_general_ci", b:[]uint8{0x30, 0x31, 0x32, 0x33}, x:interface {}(nil)} != record:types.Datum{k:0x5, decimal:0x0, length:0x0, i:0, collation:"utf8mb4_general_ci", b:[]uint8{0x30, 0x31, 0x32, 0x33, 0x34}, x:interface {}(nil)}')
```

### 4. What is your TiDB version? (Required)

release-5.0 (09a4c57d00175ae3ff9c900d5fb67f6757c58181)
br@86407e1

### PR DESCRIPTION
**Pull request by wjhuang2016**
### What problem does this PR solve?

Issue Number: close #23707 

Problem Summary:

If we create an index on the column which has been prefixed indexed by the clustered index, the restored data is truncated.
Then indexReader would read the wrong truncated data.

### What is changed and how it works?

Save the data with max length.
Says that primary key(a (8)),
For index t(a), don't truncate the value.
For index t(a(9)), truncate to a(9).
For index t(a(7)), truncate to a(8).

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test

Side effects

### Release note 

- fix a bug that wrong index data on prefixed clustered index

### CODE REVIEW

File: table/tables/tables.go [hunk 1]
```diff
@@ -1729,10 +1729,24 @@ func TryGetHandleRestoredDataWrapper(t table.Table, row []types.Datum, rowMap ma
...
-	for _, idx := range t.Meta().Indices {
-		if idx.Primary {
-			tablecodec.TruncateIndexValues(t.Meta(), idx, rsData)
-			break
+	// Try to truncate index values.
+	// Says that primary key(a (8)),
+	// For index t(a), don't truncate the value.
+	// For index t(a(9)), truncate to a(9).
+	// For index t(a(7)), truncate to a(8).
+	pkIdx := FindPrimaryIndex(t.Meta())
```
Comment (coocood) [inline table/tables/tables.go:1737]: Is it possible `pkIdx` is nil?

Comment (wjhuang2016) [reply table/tables/tables.go:1737]: No
if !collate.NewCollationEnabled() || !t.Meta().IsCommonHandle || t.Meta().CommonHandleVersion == 0 {
		return nil
	}

File: table/tables/tables.go [hunk 2]
```diff
@@ -1729,10 +1729,24 @@ func TryGetHandleRestoredDataWrapper(t table.Table, row []types.Datum, rowMap ma
...
+	for i, pkCol := range pkIdx.Columns {
+		for _, idxCol := range idx.Columns {
+			if idxCol.Offset == pkCol.Offset {
+				if idxCol.Length == types.UnspecifiedLength {
+					break
+				}
+				useIdx := idxCol
+				if pkCol.Length > idxCol.Length {
```
Comment (coocood) [inline table/tables/tables.go:1745]: What if `pkCol.Length == types.UnspecifiedLength`?

File: table/tables/tables.go [hunk 3]
```diff
@@ -1729,10 +1729,24 @@ func TryGetHandleRestoredDataWrapper(t table.Table, row []types.Datum, rowMap ma
...
+	// Says that primary key(a (8)),
+	// For index t(a), don't truncate the value.
+	// For index t(a(9)), truncate to a(9).
+	// For index t(a(7)), truncate to a(8).
+	pkIdx := FindPrimaryIndex(t.Meta())
+	for i, pkCol := range pkIdx.Columns {
+		for _, idxCol := range idx.Columns {
+			if idxCol.Offset == pkCol.Offset {
+				if idxCol.Length == types.UnspecifiedLength || pkCol.Length == types.UnspecifiedLength {
+					break
```
Comment (coocood) [inline table/tables/tables.go:1742]: We can return here.

**Review by tangenta — APPROVED** [review]

File: table/tables/tables.go [hunk 4]
```diff
@@ -1729,10 +1729,24 @@ func TryGetHandleRestoredDataWrapper(t table.Table, row []types.Datum, rowMap ma
...
 
-	for _, idx := range t.Meta().Indices {
-		if idx.Primary {
-			tablecodec.TruncateIndexValues(t.Meta(), idx, rsData)
-			break
+	// Try to truncate index values.
+	// Says that primary key(a (8)),
+	// For index t(a), don't truncate the value.
+	// For index t(a(9)), truncate to a(9).
+	// For index t(a(7)), truncate to a(8).
```
Comment (AilinKid) [inline table/tables/tables.go:1736]: better specify it in the document

Comment (wjhuang2016) [reply table/tables/tables.go:1736]: It's transparency to users.

**Pull request merged**